zebROS_ws/src/controllers_2023/src/four_bar_controller.cpp

```diff
@@ -0,0 +1,461 @@
+    },
+    [this](double b)
+    {
+        four_bar_zeroing_percent_output.store(b);
+    },
+    "FourBar Zeroing Percent Output");
+    ddr_.registerVariable<double>
+    ("four_bar_zeroing_timeout",
+     [this]()
+    {
+        return four_bar_zeroing_timeout.load();
+    },
+    [this](double b)
+    {
+        four_bar_zeroing_timeout.store(b);
+    },
+    "FourBar Zeroing Timeout");
+    ddr_.registerVariable<double>
+    ("motion_magic_velocity",
+     [this]()
+    {
+        return motion_magic_velocity.load();
+    },
+    [this](double b)
+    {
+        motion_magic_velocity.store(b);
+    },
+    "Motion Magic Velocity");
+    ddr_.registerVariable<double>
+    ("motion_magic_acceleration",
+     [this]()
+    {
+        return motion_magic_acceleration.load();
+    },
+    [this](double b)
+    {
+        motion_magic_acceleration.store(b);
+    },
+    "Motion Magic Acceleration");
+    ddr_.registerVariable<int>
+    ("motion_s_curve_strength",
+     [this]()
+    {
+        return motion_s_curve_strength.load();
+    },
+    [this](int b)
+    {
+        motion_s_curve_strength.store(b);
+    },
+    "S Curve Strength");
+    
+    ddr_.publishServicesTopics();
+
+    //initialize the four_bar joint
+    if (!four_bar_joint_.initWithNode(talon_command_iface, nullptr, controller_nh, four_bar_params))
+    {
+        ROS_ERROR("Cannot initialize four_bar joint!");
+    }
+
+    four_bar_service_ = controller_nh.advertiseService("four_bar_service", &FourBarController_2023::cmdService, this);
+
+    return true;
+}
+
+void FourBarController_2023::starting(const ros::Time &time)
+{
+    zeroed_ = false;
+    last_zeroed_  = false;
+    last_mode_ = hardware_interface::TalonMode_Disabled;
+    last_angle_ = -1; // give nonsense position to force update on first time through update()
+    position_command_.writeFromNonRT(FourBarCommand_2023(angleFromX(min_extension_, false)));
+}
+
+void FourBarController_2023::update(const ros::Time &time, const ros::Duration &/*duration*/)
+{
+    // If we hit the limit switch, (re)zero the position.
+    if (four_bar_joint_.getReverseLimitSwitch()) 
+    {
+        ROS_INFO_THROTTLE(2, "FourBarController_2023 : hit limit switch");
+        if (!last_zeroed_)
+        {
+            zeroed_ = true;
+            last_zeroed_ = true;
+            four_bar_joint_.setSelectedSensorPosition(0); // relative to min position
+            four_bar_joint_.setDemand1Type(hardware_interface::DemandType_ArbitraryFeedForward);
+            four_bar_joint_.setDemand1Value(arb_feed_forward);
+        }
+    }
+    else
+    {
+        last_zeroed_ = false;
+    }
+
+
+    if (zeroed_) // run normally, seeking to various positions
+    {
+        four_bar_joint_.setMode(hardware_interface::TalonMode_MotionMagic);
+        if (((four_bar_joint_.getMode()) == (hardware_interface::TalonMode_Disabled)) && ((last_mode_) == (hardware_interface::TalonMode_Disabled)))
```

Comment:
Lots of extra parens here - no need to put parens around a single variable name or method call, just looking for `if ((a == b) && (c == d))` kinda stuff